packages/logger/src/logger.js

```diff
@@ -144,8 +119,131 @@ export default class PercyLogger {
   }
 
-  // Query for a set of logs by filtering the in-memory store
-  query(filter) {
-    return Array.from(this.messages).filter(filter);
+  // Returns true or false if the level and debug group can write messages to stdio
+  shouldLog(debug, level) {
+    return LOG_LEVELS[level] != null &&
+      LOG_LEVELS[level] >= LOG_LEVELS[this.level] &&
+      !this.namespaces.exclude.some(ns => ns.test(debug)) &&
+      this.namespaces.include.some(ns => ns.test(debug));
+  }
+
+  // Ensures that deprecation messages are not logged more than once
+  deprecated(debug, message, meta) {
+    if (this.deprecations.has(message)) return;
+    this.deprecations.add(message);
+
+    this.log(debug, 'warn', `Warning: ${message}`, meta);
+  }
+
+  // Returns true if a socket is present and ready
+  get isRemote() {
+    return this.socket?.readyState === 1;
+  }
+
+  // Generic log method accepts a debug group, log level, log message, and optional meta
+  // information to store with the message and other info
+  log(debug, level, message, meta = {}) {
+    // message might be an error object
+    let isError = typeof message !== 'string' && (level === 'error' || level === 'debug');
+    let error = isError && message;
+
+    // if remote, send logs there
+    if (this.isRemote) {
+      // serialize error messages
+      message = isError && 'stack' in error ? {
+        message: error.message,
+        stack: error.stack
+      } : message;
+
+      return this.socket.send(JSON.stringify({
+        log: [debug, level, message, { remote: true, ...meta }]
+      }));
+    }
+
+    // ensure the message is a string
+    message = (isError && message.stack) ||
+      message.message || message.toString();
+
+    // timestamp each log
+    let timestamp = Date.now();
+    let entry = { debug, level, message, meta, timestamp };
+    this.messages.add(entry);
+
+    // maybe write the message to stdio
+    if (this.shouldLog(debug, level)) {
+      let elapsed = timestamp - (this.lastlog || timestamp);
+      if (isError && this.level !== 'debug') message = error.toString();
+      this.write(level, this.format(message, debug, error ? 'error' : level, elapsed));
+      this.lastlog ||= timestamp;
+    }
+  }
+
+  // Writes a message to stdio based on the loglevel
+  write(level, message) {
+    let stdio = level === 'info' ? 'stdout' : 'stderr';
+    this.constructor[stdio].write(message + '\n');
+  }
+
+  // Opens a socket logging connection
+  connect(socket) {
+    // send logging environment info
+    let PERCY_DEBUG = process.env.PERCY_DEBUG;
+    let PERCY_LOGLEVEL = process.env.PERCY_LOGLEVEL || this.loglevel();
+    socket.send(JSON.stringify({ env: { PERCY_DEBUG, PERCY_LOGLEVEL } }));
+
+    // attach remote logging handler and return a cleanup function
+    return listen(socket, 'message', ({ data }) => {
+      let { log, logAll } = JSON.parse(data);
+      if (logAll) logAll.forEach(e => this.messages.add(e));
```

Comment:
This will silently add previously logged remote logs to this loggers memory store. They have the potential to be added "out of order" since old logs are being added to the current Set order. However, each entry has a timestamp if there is ever confusion. If it's really important (probably not), we could potentially write a small merge util that will order them by timestamp.

Reply:
Also just had the idea of having them be sorted by the `query` method so they will be ordered in log files.